piptools/_compat/pip_compat.py

```diff
@@ -63,6 +67,19 @@ def _from_importlib(cls, dist: _ImportLibDist) -> Distribution:
         return cls(dist._dist.name, dist._dist.version, requires, dist.direct_url)
 
 
+def canonicalize_ireq(ireq: InstallRequirement) -> None:
```

Comment:
Instead of mutating the incoming object, can we make a new one and return it? Changing things passed into function is going to be confusing, reducing maintainability/readability on the calling side.

Reply:
Sure, will change it today.

Reply:
So I'd like to use `utils.copy_install_requirement`, and `utils` currently imports `PIP_VERSION` from `_compat`. 

To avoid circular imports, I suggest to move the new stuff into `utils`, import it where needed from there (including into `_compat`), and removing `utils`'s current `PIP_VERSION` import from `_compat`.

I could inline the `PIP_VERSION` logic over in `utils`, but I suggest just moving `PIP_VERSION` into `utils` altogether.

Reply:
I'm noticing we have a mix of relative and absolute imports for piptools modules, so I'm not sure which is preferred, though I like relative imports.

Reply:
Alright, please have a look at the changes and let me know what you think. Thanks!

Reply:
> I'm noticing we have a mix of relative and absolute imports for piptools modules, so I'm not sure which is preferred, though I like relative imports.

In general, I like relative imports. Though, there's cases where you want to patch things in other modules, and that might dictate how you import. Also, the context side of this is answered by https://youtu.be/x-kB2o8sd5c?si=yRWpqG3YmZRe_3AY&t=1038, I believe.

Reply:
As for the circular imports, we should probably implement https://github.com/sanitizers/octomachinery/blob/master/tests/circular_imports_test.py in pip-tools, which is obviously outside of the scope of this PR.

Reply:
By the way, one of the strategies I adopted that helps avoid surprise import-time side effects is not putting any logic into `__init__.py` files since they are evaluated when you import things levels deeper too. Of course, it's something to consider separately.

Reply:
Looking good in this regard?

Reply:
Marking this as resolved.